### ISSUE TITLE
proposal: disallow `extends` and `mixins`

### ISSUE THREAD

**Issue by privatenumber** [vuejs/eslint-plugin-vue#1315]
**Please describe what the rule should do:**
Warn or error on the use of object-oriented/inheritance paradigms like `extends` and `mixins` as a "Strongly Recommended" rule.

This can currently be implemented on a individual level by configuring [`vue/no-restricted-component-options`](https://eslint.vuejs.org/rules/no-restricted-component-options.html), but by making it a recommended rule in the plugin, it could help beginners from object-oriented backgrounds consider a composition approach by preventing them from shoehorning OO implementations into a composition framework.

| | `extends` | `mixins` |
| - | - | - |
| Vue 2 | Error | Warn |
| Vue 3 | Error | Error (in favor of [Composition API](https://composition-api.vuejs.org/)) |

**What category should the rule belong to?**

- [x] Enforces code style (layout)
- [ ] Warns about a potential error (problem)
- [x] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

Sorry I checked two. I'm indifferent about which approach to take. Maybe `suggestion` for Vue 2 and `layout` for Vue 3?

If `suggestion`, the rule should suggest that a composition approach be taken instead.

**Provide 2-3 code examples that this rule should warn about:**

**`extends`**
```vue
<!-- No template tag because it's inherited -->

<script>
import ComponentC from './ComponentC'; // which could be extending ComponentB and ComponentA

export default {
	extends: ComponentC,

	data() {
		...
	},

	...
};
</script>
```

**`mixins`**
```vue
<script>
const MixinA = {
	methods: {
		doThis() { ... }
	}
};

const MixinB = {
	methods: {
		doThis() { ... }
	}
};

export default {
	mixins: [
		// Colliding mixins
		MixinA,
		MixinB,
		...
	],

	data() {
		...
	},

	...
};
</script>
```

**Additional context**

`mixins`/`extends` can be over-used to a point where the reader must look at multiple files to understand what's happening in one component. Looking at multiple files and merging them in your head requires a higher mental load and goes against the encapsulation benefit single-file components (SFC) promotes.

React's Dan Abramov dives deeper into why _inheritance_ is an anti-pattern to _composition_ in his 2016 blog post ["Mixins Considered Harmful"](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html).

He explains that, in React, `mixins` were added to ease the initial adoption and learning curve of a composition approach because at the time, the majority of JS UI libraries were object-oriented. (I imagine Vue has a similar reason too.)

His points:
- [**Implicit dependencies**](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html#mixins-introduce-implicit-dependencies) `mixins` or `extends` can define properties & methods on a component that cannot be effectively enforced or documented.
- [**Collisions**](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html#mixins-cause-name-clashes) There's no guarantee that two mixins can be mixed together because they can be using the same property/method names. In Vue, there are no warnings when a collision happens at run-time or build-time (eg. [Extends](https://jsfiddle.net/hirokiosame/pyd9z8ow/) / [Mixins](https://jsfiddle.net/hirokiosame/nfo152ae/)).
- [**Snowballing complexity**](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html#mixins-cause-snowballing-complexity) Mixins grow complex over time and become increasingly coupled with the component using them. (Applies mainly to in-house mixins/extends and not 3rd party ones.)
- **Bad for static analysis** This reason isn't explicitly given, but it's a common thread across the problems. I'm sure this is very relevant to `eslint-plugin-vue` because components that use `extends`/`mixins` can't be linted accurately.

Not all `mixins`/`extends` are bad, but can be over-used to a point where you're losing maintainability and understanding of what's happening in your code.

A linting rule to catch this practice early-on and suggest that a composition approach be considered instead would be very helpful, especially for those coming in from an OOP background.

**Comment by ota-meshi** [comment]
Thank you for this issue.

However, can't you use the [no-restricted-component-options] rule to meet your request?

[no-restricted-component-options]: https://eslint.vuejs.org/rules/no-restricted-component-options.html

**Comment by privatenumber** [comment]
Thanks for pointing that out @ota-meshi 

I updated the proposal to take that into consideration; essentially, I think having a "recommended" rule that guides beginners from OO backgrounds into a composition approach would be beneficial to the community.

I've seen some pretty gnarly use of `extends` that goes ~4 levels deep, with each file being 200+ lines. There's currently no good Vue guide/tooling that helps re-think OO patterns into composition ones, and I'm thinking `eslint-plugin-vue` would be a great place to start.

Interested in hearing your opinion on using `mixins`/`extends` in Vue if you're open to sharing.

**Comment by ota-meshi** [comment]
I think it's bad to report `mixins` and `extends` with the recommended configurations provided by `eslint-plugin-vue`.
If the frameworks used by users can only be used by using `mixins`, they can't do anything.
If you're creating a framework that recommends not using `mixins`, I think you should provide an ESLint sharable configuration for that your framework users. You can use the [vue/no-restricted-component-options] rule in that configuration.

For example, Vuetify is creating an eslint-plugin for Vuetify. 
https://github.com/vuetifyjs/eslint-plugin-vuetify
If you are using Vuetify, you may need to suggest a rule to report `mixins` and `extends`, to `eslint-plugin-vuetify`.

[vue/no-restricted-component-options]: https://eslint.vuejs.org/rules/no-restricted-component-options.html
